src/GhciFind.hs

```diff
@@ -49,12 +83,30 @@ findCompletions infos fp sample sl sc el ec =
              return (Left ("No module info for the current file! Try loading it?"))
            Just moduleInf ->
              do df <- getDynFlags
-                let toplevelNames =
-                      fromMaybe [] (modInfoTopLevelScope (modinfoInfo moduleInf))
+                (qual, ident, minfs) <-
+                   let noQual = ("", sample, [modinfoInfo moduleInf])
+                       getModInfo qmname =
+                         findModule qmname Nothing >>= getModuleInfo
+                   in if '.' `elem` sample
+                      then do let importDecls = importDeclsFromInfo moduleInf
+                              case findQualifiedSource
+                                      importDecls
+                                      sample of
+                                 Just (qual, ident, qualModNames) -> do
+                                   minfos <- fmap catMaybes
+                                                  (mapM getModInfo qualModNames)
+                                   if null minfos
+                                      then return noQual
+                                      else return (qual, ident, minfos)
+                                 Nothing -> return noQual
+                      else return noQual
+                let toplevelNames = concat (mapMaybe modInfoTopLevelScope minfs)
                     filteredToplevels =
-                      filter (isPrefixOf sample)
```

Comment:
These are good ideas. I think we should also keep a very close eye on performance as this can be a very big constant factor on completion generation time in the best case. I think pre-filtering possible module sources as is done helps, but it would be great if we had a way of measuring things in a somewhat reliable way. We could put a timer on the elisp side and log that when in a debug mode.